src/client/datascience/telemetry/telemetry.ts

```diff
@@ -55,11 +58,17 @@ export function sendKernelTelemetryEvent<P extends IEventNamePropertyMapping, E
         setSharedProperty('userExecutedCell', 'true');
     }
 
-    const addOnTelemetry = getContextualPropsForTelemetry(resource);
-    const props = properties || {};
-    Object.assign(props, addOnTelemetry || {});
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    sendTelemetryEvent(eventName as any, durationMs, props, ex, true);
+    const props = getContextualPropsForTelemetry(resource);
```

Comment:
I had another thought. Would it be possible to send properties as promises? Then in the sendTelemetryEvent just check if any properties were promises.

If any were promises, put the event into a queue that submits once all of promises finish
If any event comes in, and the queue has an item, it goes into the queue too.

Reply:
The problem is with our strong typing, we only allow sending telemetry properties that are known.
Her'es a sample 
```typescript
    [Telemetry.KernelEnumeration]: {
        /**
         * Count of the number of kernels found
         */
        count: number;
        /**
         * Boolean indicating if any are python or not
         */
        isPython: boolean;
        /**
         * Indicates how the enumeration was acquired.
         */
        source: 'cli' | 'connection';
    };
```

Here' we'll now need to add a property with a promise.
I.e. we'll need to update `IEventNamePropertyMapping` to contain a new promise.
The current approach makes it unnecessary to update `IEventNamePropertyMapping`

> If any event comes in, and the queue has an item, it goes into the queue too.

This already happens today, i just check if the queue is empty or not.

Reply:
Wouldn't that be better? Having strong typing for promises? Not sure why that's a downside.

I need to look some more but your current implementation is passing a flag to indicate if should be queued or not?

Reply:
& now we'll have 
```
    [Telemetry.KernelEnumeration]: {
        /**
         * Count of the number of kernels found
         */
        count: number;
        /**
         * Boolean indicating if any are python or not
         */
        isPython: boolean;
        /**
         * Indicates how the enumeration was acquired.
         */
        source: 'cli' | 'connection';
		// New Property.
		isReady: Promise<void>
    };
```

Optionally, we could allows the property `isReady` to all telemetry properties passed into the `sendTelemetryEvent` (& IEventNamePropertyMapping is left untouched) & when we send the telemetry we jsut drop that property.

How does that sound (pretty much what you are suggesting)?

Reply:
Sure that works too.